src/bootloaders/shim-systemd.c

```diff
@@ -0,0 +1,313 @@
+{
+        (void)manager;
+        return strdup(KERNEL_DST_DIR);
+}
+
+static bool shim_systemd_install_kernel(const BootManager *manager, const Kernel *kernel)
+{
+        return sd_class_install_kernel_impl(manager, kernel, shim_systemd_get_kernel_dst, NULL);
+}
+
+static bool shim_systemd_remove_kernel(const BootManager *manager, const Kernel *kernel)
+{
+        return sd_class_remove_kernel(manager, kernel);
+}
+
+static bool shim_systemd_set_default_kernel(const BootManager *manager, const Kernel *kernel)
+{
+        /* this writes systemd config. systemd has the configuration paths
+         * hardcoded, hence whatever sd_class is doing is OK. */
+        return sd_class_set_default_kernel_impl(manager, kernel, NULL);
+}
+
+static bool exists_identical(const char *path, const char *spath)
+{
+        if (!nc_file_exists(path))
+                return false;
+        if (spath && !cbm_files_match(path, spath))
+                return false;
+        return true;
+}
+
+static bool shim_systemd_needs_install(const BootManager *manager)
+{
+        (void)manager;
+        if (!exists_identical(shim_dst_host, NULL))
+                return true;
+        if (!exists_identical(systemd_dst_host, NULL))
+                return true;
+        return false;
+}
+
+static bool shim_systemd_needs_update(const BootManager *manager)
+{
+        (void)manager;
+        if (!exists_identical(shim_dst_host, shim_src))
+                return true;
+        if (!exists_identical(systemd_dst_host, systemd_src))
+                return true;
+        return false;
+}
+
+static bool make_layout(const BootManager *manager)
+{
+        char *boot_root = boot_manager_get_boot_dir((BootManager *)manager);
+        char path[PATH_MAX];
+        snprintf(path, PATH_MAX, "%s%s", boot_root, DST_DIR);
+        if (!nc_mkdir_p(path, 00755)) {
+                goto fail;
+        }
+        snprintf(path, PATH_MAX, "%s%s", boot_root, KERNEL_DST_DIR);
+        if (!nc_mkdir_p(path, 00755)) {
+                goto fail;
+        }
+        snprintf(path, PATH_MAX, "%s%s", boot_root, SYSTEMD_ENTRIES);
+        if (!nc_mkdir_p(path, 00755)) {
+                goto fail;
+        }
+        return true;
+fail:
+        LOG_FATAL("Failed to make dir: %s", path);
+        return false;
+}
+
+static bool shim_systemd_install(const BootManager *manager)
+{
+        char varname[9];
+
+        if (!make_layout(manager)) {
+                LOG_FATAL("Cannot create layout");
+                return false;
+        }
+
+        if (!copy_file_atomic(shim_src, shim_dst_host, 00644)) {
+                LOG_FATAL("Cannot copy %s to %s", shim_src, shim_dst_host);
+                return false;
+        }
+        if (!copy_file_atomic(systemd_src, systemd_dst_host, 00644)) {
+                LOG_FATAL("Cannot copy %s to %s", systemd_src, systemd_dst_host);
+                return false;
+        }
+
+        if (bootvar_create(BOOT_DIRECTORY, shim_dst_esp, varname, 9)) {
```

Comment:
It might be more pertinent to make this ERROR not FATAL, as dodgy firmware implementations are just going to fail here if there is no room left for new vars

Reply:
if the EFI var cannot be created, then it's a fatal error for this boot scheme.